### PR TITLE
Retry with backoff when rate-limited

### LINKED ISSUE #32: Built-in retry with exponential back-offs
It would be nice if the client had a built-in retry + back-off algorithm. Not only is this a great feature to add for the general case, it would also be super helpful in the Terraform Heroku provider (see terraform-providers/terraform-provider-heroku/issues/71 for more background.)

I'm happy to work on an implementation, lets discuss first though.

We could look at bringing in https://github.com/cenkalti/backoff or lifting the logic from elsewhere.

### PR DESCRIPTION
**Pull request by mars**
### Context

Heroku Platform API:

* [rate limits to 4500/requests/hour/account](https://devcenter.heroku.com/articles/platform-api-reference#rate-limits)
* sends current usage count in an HTTP response header:
  
  ```
  Ratelimit-Remaining: 4499
  ```
* returns [HTTP status **429**](https://devcenter.heroku.com/articles/platform-api-reference#client-error-responses) when the account is rate limited.

### Proposal

Let's make this client real friendly with HTTP status **429** response errors 🤗 to improve durability of the [Terraform Provider](https://github.com/terraform-providers/terraform-provider-heroku).

Implement retry with exponential backoff when rate-limited with a new `RoundTripWithRetryBackoff` transport wrapping the default net/http `RoundTrip()`.

Proposed (& now implemented) [usage](https://github.com/terraform-providers/terraform-provider-heroku/blob/master/heroku/config.go#L24), note the `Transport` field:

```go
heroku.NewService(&http.Client{
  Transport: &heroku.Transport{
    Username:          c.Email,
    Password:          c.APIKey,
    UserAgent:         heroku.DefaultUserAgent,
    AdditionalHeaders: c.Headers,
    Debug:             debugHTTP,
    Transport:         heroku.RoundTripWithRetryBackoff{
      // These are default values for ExponentialBackOff config
      InitialIntervalSeconds: 30,
      RandomizationFactor:    0.25,
      Multiplier:             2,
      MaxIntervalSeconds:     900,
      MaxElapsedTimeSeconds:  0,
    },
  },
})
```

The client reacts to the terminal HTTPS status **429** by retrying with back-off, instead of trying to predict and back-off before being rate-limited. Using the `Ratelimit-Remaining` header intelligently would seem to require global state shared across [Terraform's parallel execution of nodes](https://www.terraform.io/docs/internals/graph.html#walking-the-graph), so this proposal avoids that seemingly unnecessary complexity.

Fixes #32

### CODE REVIEW

File: v3/round_trip_with_retry_backoff.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+	"github.com/cenkalti/backoff"
+)
+
+func RoundTripWithRetryBackoff(req *http.Request) (*http.Response, error) {
+	var lastResponse *http.Response
+	var lastError error
+
+	retryableRoundTrip := func() error {
+		lastResponse, lastError = http.DefaultTransport.RoundTrip(req)
```
Comment (sclasen) [inline v3/round_trip_with_retry_backoff.go]: probably need to check lastError and return it if not nol?

Comment (mars) [reply v3/round_trip_with_retry_backoff.go]: `lastError` is checked just a few lines down [in the parent function](#diff-5c57c0fa1ac47d01d237b2e53639943aR33), because this `backoff.Retry` function should only return an error when it should be retried (status 429 for this rate-limiting use-case).

Comment (talbright) [reply v3/round_trip_with_retry_backoff.go]: You could dry it up by dropping `lastError` and doing something like this:

```
retryableRoundTrip := func() err error {
		lastResponse, err = http.DefaultTransport.RoundTrip(req)
		// Detect Heroku API rate limiting
		// https://devcenter.heroku.com/articles/platform-api-reference#client-error-responses
		if lastResponse.StatusCode == 429 {
			err = fmt.Errorf("Heroku API rate limited: 429 Too Many Requests")
		}
		return err
	}
```

That would also mean that you would only need to handle the error once below...

Comment (mars) [reply v3/round_trip_with_retry_backoff.go]: But we only want this to retry for HTTP status **429** responses.

If that function returns other errors, it will retry for them too.

We only want to retry when rate-limited, right?

Comment (talbright) [reply v3/round_trip_with_retry_backoff.go]: Oh yeah 😊

**Comment by mars** [comment]
I've verified that this retrier is doing the right thing by switching to retry on to HTTP status 200, and then watch the logs during tests with `TF_LOG=DEBUG` to see it retrying.

**Comment by mars** [comment]
To test I revised the `backoff.RetryNotify` function:

```go
retryableRoundTrip := func() error {
	lastResponse, lastError = http.DefaultTransport.RoundTrip(req)
	// Detect Heroku API rate limiting
	// https://devcenter.heroku.com/articles/platform-api-reference#client-error-responses
	if lastResponse.StatusCode == 429 {
		return fmt.Errorf("Heroku API rate limited: 429 Too Many Requests")
	}

	// Fake retry testing.
	if lastResponse.StatusCode == 200 {
		return fmt.Errorf("Fake retrier!!!")
	}

	return nil
}
```

…and can see the randomized backoff working in the logs:

```
2018/10/25 09:14:00 Will retry Heroku API request in 33.872687193s, because Fake retrier!!!
2018/10/25 09:14:35 Will retry Heroku API request in 46.371947138s, because Fake retrier!!!
2018/10/25 09:15:21 Will retry Heroku API request in 1m46.253268969s, because Fake retrier!!!
2018/10/25 09:17:08 Will retry Heroku API request in 4m0.01901876s, because Fake retrier!!!
2018/10/25 09:21:09 Will retry Heroku API request in 8m21.662152299s, because Fake retrier!!!
```

So, my confidence is building that this solution is good 😇

**Comment by mars** [comment]
I think this is ready to go!

The new retry functionality is not enabled by default. `Transport: heroku.RoundTripWithRetryBackoff{}` must be specifically set. So there's very little chance of this affecting existing usage.

Please review and let me know what you think.

File: v3/round_trip_with_retry_backoff.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/cenkalti/backoff"
+)
+
+// net/http RoundTripper interface, a.k.a. Transport
+// https://godoc.org/net/http#RoundTripper
+type RoundTripWithRetryBackoff struct {
```
Comment (talbright) [inline v3/round_trip_with_retry_backoff.go]: ✅ good call -- was going to suggest in an earlier commit you allow some configuration to be passed through

File: v3/round_trip_with_retry_backoff.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+	rateLimitRetryConfig := &backoff.ExponentialBackOff{
+		Clock:               backoff.SystemClock,
+		InitialInterval:     time.Duration(int64WithDefault(r.InitialIntervalSeconds, int64(30))) * time.Second,
+		RandomizationFactor: float64WithDefault(r.RandomizationFactor, float64(0.25)),
+		Multiplier:          float64WithDefault(r.Multiplier, float64(2)),
+		MaxInterval:         time.Duration(int64WithDefault(r.MaxIntervalSeconds, int64(900))) * time.Second,
+		MaxElapsedTime:      time.Duration(int64WithDefault(r.MaxElapsedTimeSeconds, int64(0))) * time.Second,
+	}
+	rateLimitRetryConfig.Reset()
```
Comment (talbright) [inline v3/round_trip_with_retry_backoff.go]: Why did you need to do this? (https://godoc.org/github.com/cenkalti/backoff#ExponentialBackOff.Reset)

Comment (mars) [reply v3/round_trip_with_retry_backoff.go]: I followed the way the backoff module's built-in convenience constructor works:
https://github.com/cenkalti/backoff/blob/adb73d5bf0d9237fab19ff58aebf658449e326df/exponential.go#L92

Not sure it's required, but it seemed prudent to follow the creators lead 😅

Comment (talbright) [reply v3/round_trip_with_retry_backoff.go]: That sounds logical to me 👍

Comment (talbright) [reply v3/round_trip_with_retry_backoff.go]: https://github.com/cenkalti/backoff/blob/adb73d5bf0d9237fab19ff58aebf658449e326df/exponential.go#L107-L108

**Comment by mars** [comment]
Thanks @talbright 😄

**Comment by schneems** [comment]
Hello rate throttlers! I've been poking at this problem of adding rate throttling to the `platform-api` gem for a bit. 

I did some benchmarking and it looks like if one of the goals is to reduce the total number of requests to the origin server that an exponential backoff strategy doesn't help that much. In my benchmark it has an 80% retry rate. If the primary goal is to not have failures due to rate limiting, then it does a great job there. Here's my benchmark results:

https://github.com/zombocom/rate_throttle_client#ratethrottleclientexponentialbackoff-results-duration-300-minutes-multiplier-12

And here's some of my research and how I went about iterating on a solution with good properties that has a lower retry rate (~3% instead of 80%): https://github.com/schneems/rate_throttle_clients#how-to-write-a-rate-throttling-algorithm
 
Here's the `platform-api` PR https://github.com/heroku/platform-api/pull/103. 

There's no immediate action needed by y'all but if you're interested in various rate-throttling strategies, it's an active area of research/interest of mine.

**Comment by mars** [comment]
So thoughtful @schneems 🙌 As you intuited, the reason rate-throttling was implemented here was to avoid `terraform apply` failures.
